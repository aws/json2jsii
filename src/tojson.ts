import { Code } from './code';

export class toJsonFunction {
  public static nameOf(typeName: string) {
    return `${typeName}$toJson`;
  }

  private readonly fields: Record<string, string> = {};

  public readonly functionName: string;

  constructor(private readonly baseType: string) {
    this.functionName = toJsonFunction.nameOf(this.baseType);
  }

  public addField(schemaName: string, propertyName: string, toJson: toJson) {
    this.fields[schemaName] = toJson(`obj.${propertyName}`);
  }

  public emit(code: Code) {
    code.line();
    code.line('/**');
    code.line(` * Converts an object of type '${this.baseType}' to JSON representation.`);
    code.line(' */');
    code.openBlock(`export function ${this.functionName}(obj: ${this.baseType} | undefined): Record<string, any> | undefined`);
    code.line('if (obj === undefined) { return undefined; }');
    code.open('const result = {');
    for (const [k, v] of Object.entries(this.fields)) {
      code.line(`'${k}': ${v},`);
    }
    code.close('};');

    code.line('// filter undefined values');
    code.line('return Object.entries(result).reduce((r, i) => (i[1] === undefined) ? r : ({ ...r, [i[0]]: i[1] }), {});');

    code.closeBlock();
  }
}

export type toJson = (x: string) => string;