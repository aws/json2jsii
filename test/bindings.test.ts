import * as fs from 'fs';
import * as os from 'os';
import * as path from 'path';
import { srcmak } from 'jsii-srcmak';
import { TypeGenerator } from '../src';

jest.setTimeout(5 * 60 * 1000);

test('language bindings', async () => {
  const g = new TypeGenerator();

  g.addType('Name', {
    properties: {
      first: { type: 'string' },
      middle: { type: 'string' },
      last: { type: 'string' },
      objectWithTypedAdditionalPropertiesOnly: {
        type: 'object',
        additionalProperties: {
          type: 'string',
        },
      },
      objectWithPropertiesAndBooleanAdditionalProperties: {
        type: 'object',
        properties: {
          name: { type: 'string' },
        },
        additionalProperties: {},
      },
      objectWithPropertiesAndTypedAdditionalProperties: {
        type: 'object',
        properties: {
          name: { type: 'string' },
        },
        additionalProperties: {
          type: 'string',
        },
      },
    },
    additionalProperties: true,
    required: ['first', 'last'],
  });

  const workdir = fs.mkdtempSync(path.join(os.tmpdir(), 'json2jsii'));

  const src = path.join(workdir, 'tyepscript');
  fs.mkdirSync(src);
  fs.writeFileSync(path.join(src, 'index.ts'), await g.render());

  await srcmak(src, {
    java: {
      outdir: path.join(workdir, 'java'),
      package: 'org.myorg',
    },
    python: {
      outdir: path.join(workdir, 'python'),
      moduleName: 'myorg',
    },
    golang: {
      outdir: path.join(workdir, 'golang'),
      packageName: 'myorg',
      moduleName: 'myorg',
    },
    csharp: {
      outdir: path.join(workdir, 'csharp'),
      namespace: 'myorg',
    },
  });

  expect(readFile(path.join(workdir, 'python/myorg/__init__.py'))).toMatchSnapshot();
  expect(readFile(path.join(workdir, 'java/src/main/java/org/myorg/Name.java'), ['@javax.annotation.Generated'])).toMatchSnapshot();
  expect(readFile(path.join(workdir, 'golang/myorg/myorg.go'))).toMatchSnapshot();
  expect(readFile(path.join(workdir, 'csharp/myorg/myorg/IName.cs'))).toMatchSnapshot();
});

function readFile(filePath: string, ignoreLines: string[] = []) {
  const lines = (fs.readFileSync(filePath, 'utf-8')).split('\n');
  const shouldInclude = (line: string) => !ignoreLines.find(pattern => line.includes(pattern));
  return lines.filter(shouldInclude).join('\n');
}
