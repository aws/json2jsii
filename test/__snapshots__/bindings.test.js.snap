// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`language bindings 1`] = `
"import abc
import builtins
import datetime
import enum
import typing

import jsii
import jsii.compat
import publication

from ._jsii import *


@jsii.data_type(jsii_type=\\"generated.Name\\", jsii_struct_bases=[], name_mapping={'first': 'first', 'last': 'last', 'middle': 'middle'})
class Name():
    def __init__(self, *, first: str, last: str, middle: typing.Optional[str]=None) -> None:
        \\"\\"\\"
        :param first: 
        :param last: 
        :param middle: 

        schema:
        :schema:: Name
        \\"\\"\\"
        self._values = {
            'first': first,
            'last': last,
        }
        if middle is not None: self._values[\\"middle\\"] = middle

    @builtins.property
    def first(self) -> str:
        \\"\\"\\"
        schema:
        :schema:: Name#first
        \\"\\"\\"
        return self._values.get('first')

    @builtins.property
    def last(self) -> str:
        \\"\\"\\"
        schema:
        :schema:: Name#last
        \\"\\"\\"
        return self._values.get('last')

    @builtins.property
    def middle(self) -> typing.Optional[str]:
        \\"\\"\\"
        schema:
        :schema:: Name#middle
        \\"\\"\\"
        return self._values.get('middle')

    def __eq__(self, rhs) -> bool:
        return isinstance(rhs, self.__class__) and rhs._values == self._values

    def __ne__(self, rhs) -> bool:
        return not (rhs == self)

    def __repr__(self) -> str:
        return 'Name(%s)' % ', '.join(k + '=' + repr(v) for k, v in self._values.items())


__all__ = [
    \\"Name\\",
]

publication.publish()
"
`;

exports[`language bindings 2`] = `
"package org.myorg;

/**
 */
@software.amazon.jsii.Jsii(module = org.myorg.$Module.class, fqn = \\"generated.Name\\")
@software.amazon.jsii.Jsii.Proxy(Name.Jsii$Proxy.class)
public interface Name extends software.amazon.jsii.JsiiSerializable {

    /**
     */
    @org.jetbrains.annotations.NotNull java.lang.String getFirst();

    /**
     */
    @org.jetbrains.annotations.NotNull java.lang.String getLast();

    /**
     */
    default @org.jetbrains.annotations.Nullable java.lang.String getMiddle() {
        return null;
    }

    /**
     * @return a {@link Builder} of {@link Name}
     */
    static Builder builder() {
        return new Builder();
    }
    /**
     * A builder for {@link Name}
     */
    public static final class Builder implements software.amazon.jsii.Builder<Name> {
        private java.lang.String first;
        private java.lang.String last;
        private java.lang.String middle;

        /**
         * Sets the value of {@link Name#getFirst}
         * @param first the value to be set. This parameter is required.
         * @return {@code this}
         */
        public Builder first(java.lang.String first) {
            this.first = first;
            return this;
        }

        /**
         * Sets the value of {@link Name#getLast}
         * @param last the value to be set. This parameter is required.
         * @return {@code this}
         */
        public Builder last(java.lang.String last) {
            this.last = last;
            return this;
        }

        /**
         * Sets the value of {@link Name#getMiddle}
         * @param middle the value to be set.
         * @return {@code this}
         */
        public Builder middle(java.lang.String middle) {
            this.middle = middle;
            return this;
        }

        /**
         * Builds the configured instance.
         * @return a new instance of {@link Name}
         * @throws NullPointerException if any required attribute was not provided
         */
        @Override
        public Name build() {
            return new Jsii$Proxy(first, last, middle);
        }
    }

    /**
     * An implementation for {@link Name}
     */
    final class Jsii$Proxy extends software.amazon.jsii.JsiiObject implements Name {
        private final java.lang.String first;
        private final java.lang.String last;
        private final java.lang.String middle;

        /**
         * Constructor that initializes the object based on values retrieved from the JsiiObject.
         * @param objRef Reference to the JSII managed object.
         */
        protected Jsii$Proxy(final software.amazon.jsii.JsiiObjectRef objRef) {
            super(objRef);
            this.first = this.jsiiGet(\\"first\\", java.lang.String.class);
            this.last = this.jsiiGet(\\"last\\", java.lang.String.class);
            this.middle = this.jsiiGet(\\"middle\\", java.lang.String.class);
        }

        /**
         * Constructor that initializes the object based on literal property values passed by the {@link Builder}.
         */
        private Jsii$Proxy(final java.lang.String first, final java.lang.String last, final java.lang.String middle) {
            super(software.amazon.jsii.JsiiObject.InitializationMode.JSII);
            this.first = java.util.Objects.requireNonNull(first, \\"first is required\\");
            this.last = java.util.Objects.requireNonNull(last, \\"last is required\\");
            this.middle = middle;
        }

        @Override
        public java.lang.String getFirst() {
            return this.first;
        }

        @Override
        public java.lang.String getLast() {
            return this.last;
        }

        @Override
        public java.lang.String getMiddle() {
            return this.middle;
        }

        @Override
        public com.fasterxml.jackson.databind.JsonNode $jsii$toJson() {
            final com.fasterxml.jackson.databind.ObjectMapper om = software.amazon.jsii.JsiiObjectMapper.INSTANCE;
            final com.fasterxml.jackson.databind.node.ObjectNode data = com.fasterxml.jackson.databind.node.JsonNodeFactory.instance.objectNode();

            data.set(\\"first\\", om.valueToTree(this.getFirst()));
            data.set(\\"last\\", om.valueToTree(this.getLast()));
            if (this.getMiddle() != null) {
                data.set(\\"middle\\", om.valueToTree(this.getMiddle()));
            }

            final com.fasterxml.jackson.databind.node.ObjectNode struct = com.fasterxml.jackson.databind.node.JsonNodeFactory.instance.objectNode();
            struct.set(\\"fqn\\", om.valueToTree(\\"generated.Name\\"));
            struct.set(\\"data\\", data);

            final com.fasterxml.jackson.databind.node.ObjectNode obj = com.fasterxml.jackson.databind.node.JsonNodeFactory.instance.objectNode();
            obj.set(\\"$jsii.struct\\", struct);

            return obj;
        }

        @Override
        public boolean equals(Object o) {
            if (this == o) return true;
            if (o == null || getClass() != o.getClass()) return false;

            Name.Jsii$Proxy that = (Name.Jsii$Proxy) o;

            if (!first.equals(that.first)) return false;
            if (!last.equals(that.last)) return false;
            return this.middle != null ? this.middle.equals(that.middle) : that.middle == null;
        }

        @Override
        public int hashCode() {
            int result = this.first.hashCode();
            result = 31 * result + (this.last.hashCode());
            result = 31 * result + (this.middle != null ? this.middle.hashCode() : 0);
            return result;
        }
    }
}
"
`;
